### PR TITLE
Add public IsInitOnly API and use it to fix code generation of 'init' 

### PR DESCRIPTION
**Pull request by jcouv**
Question on compiler API:
- Should the set accessor of an invalid property `static int Property { get; init; }` have `IsInitOnly` true or false? Currently, it is false, to be consistent with behavior when loading such property from PE.

IDE scenarios supported by this API: generate override, generate implementation, view metadata-as-source.

Relates to https://github.com/dotnet/roslyn/issues/40726 (test plan for records/init-only)

### CODE REVIEW

**Comment by gafter** [comment]
`initonly` is an existing concept in ECMA-335.  Unless this represents the same concept, please rename it.

**Comment by jcouv** [comment]
Tagging @jaredpar since any decision we make for public API should preferably align with the language spec

> initonly is an existing concept in ECMA-335. Unless this represents the same concept, please rename it.

Yes, the runtime has the concept of `initonly`, but C# only has the concept of `readonly` (ECMA-334 only mentions `readonly`, not `initonly`). It is sad that the names don't align, but I don't think it precludes the best name available at the C# level for this feature.

FWIW, I also considered `IsInit` and `IsExternalInit` (to match the modreq type). But "init-only" is the way LDM and the spec described such members (we had an explicit discussion about why the "only" part is useful/meaningful). `init` accessors were thought of as a shorthand for `initonly set`, and we may still introduce the `initonly` modifier methods (question is queued).

Suggestions welcome. 
I will gladly leave a question open or PROTOTYPE marker to move forward.

File: src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs [hunk 1]
```diff
@@ -1986,6 +2064,111 @@ public class C
...
+
+            var eventSymbol = comp.GlobalNamespace.GetMember<EventSymbol>("C.Event");
+            Assert.False(eventSymbol.AddMethod.IsInitOnly);
+            Assert.False(eventSymbol.GetPublicSymbol().AddMethod.IsInitOnly);
+            Assert.False(eventSymbol.RemoveMethod.IsInitOnly);
+            Assert.False(eventSymbol.GetPublicSymbol().RemoveMethod.IsInitOnly);
+        }
+
+        [Fact]
+        public void ConstructorAndDestructorAreNotInitOnly()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InitOnlyMemberTests.cs:2091]: >ConstructorAndDestructorAreNotInitOnly [](start = 20, length = 38)

Could also cover operators

File: src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs [hunk 2]
```diff
@@ -120,6 +121,39 @@ class C
...
+@"interface I
+{
+    object Property { get; init; }
+}";
+            var compilation = CreateCompilation(source);
+
+            var sourceModule = compilation.SourceModule;
+            var sourceAssembly = (SourceAssemblySymbol)sourceModule.ContainingAssembly;
+
+            var retargetingAssembly = new RetargetingAssemblySymbol(sourceAssembly, isLinked: false);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs:137]: >var retargetingAssembly = new RetargetingAssemblySymbol(sourceAssembly, isLinked: false); [](start = 12, length = 89)

Please do not use this legacy way of testing retargeting. Create a scenario when the system itself creates one, a real scenario and test that.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs [hunk 3]
```diff
@@ -120,6 +121,39 @@ class C
             RetargetingSymbolChecker.CheckSymbols(sourceNamespace.GetMember<NamedTypeSymbol>("C"), retargetingNamespace.GetMember<NamedTypeSymbol>("C"));
         }
 
+        [Fact, CompilerTrait(CompilerFeature.InitOnlySetters)]
+        public void RetargetProperties_WithInitOnlySetter()
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs:125]: >RetargetProperties_WithInitOnlySetter [](start = 20, length = 37)

I suggest that we keep this test next to other InitOnly tests.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 3)

File: src/EditorFeatures/VisualBasic/NavigationBar/GenerateFinalizerItem.vb [hunk 4]
```diff
@@ -48,6 +48,7 @@ Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.NavigationBar
                 modifiers:=New DeclarationModifiers(isOverride:=True),
                 returnType:=compilation.GetSpecialType(SpecialType.System_Void),
                 refKind:=RefKind.None,
+                isInitOnly:=False,
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/VisualBasic/NavigationBar/GenerateFinalizerItem.vb:51]: this feels weird.  how do we do other accessors?  isn't this just a certain type of accessor, not a method?

File: src/Features/CSharp/Portable/CodeRefactorings/ConvertLocalFunctionToMethod/CSharpConvertLocalFunctionToMethodCodeRefactoringProvider.cs [hunk 5]
```diff
@@ -114,6 +114,7 @@ public override async Task ComputeRefactoringsAsync(CodeRefactoringContext conte
                 modifiers: new DeclarationModifiers(isStatic, isAsync: declaredSymbol.IsAsync),
                 returnType: declaredSymbol.ReturnType,
                 refKind: default,
+                isInitOnly: false,
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/CodeRefactorings/ConvertLocalFunctionToMethod/CSharpConvertLocalFunctionToMethodCodeRefactoringProvider.cs:117]: not really getting how methods are init-only.

**Comment by CyrusNajmabadi** [comment]
done with IDE pass.  very large question about the design here.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs (hunk 2 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs:137]: I've looked at more recent retargeting tests (Fred's recent PRs) but could not figure out the key to triggering retargeting logic to kick in. I tried tracing it back (seems to be triggered in ReferenceManager?).
What's a simple/modern test you'd recommend as a model for retargeting?

---
In reply to: [422534751](https://github.com/dotnet/roslyn/pull/44077#discussion_r422534751) [](ancestors = 422534751)

File: src/Features/CSharp/Portable/CodeRefactorings/ConvertLocalFunctionToMethod/CSharpConvertLocalFunctionToMethodCodeRefactoringProvider.cs (hunk 5 above)
Comment (jcouv) [reply src/Features/CSharp/Portable/CodeRefactorings/ConvertLocalFunctionToMethod/CSharpConvertLocalFunctionToMethodCodeRefactoringProvider.cs:117]: Currently, only accessors can be init-only. But accessors are methods.
We have a discussion queued to discuss init-only methods more generally, but that should not affect this PR (assume that we don't have generalized init-only methods).

I can make the parameter optional, so it's not in your face at most places where it could not legitimately be set to true. Would that resolve your concern?

---
In reply to: [422535858](https://github.com/dotnet/roslyn/pull/44077#discussion_r422535858) [](ancestors = 422535858)

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/CodeRefactorings/ConvertLocalFunctionToMethod/CSharpConvertLocalFunctionToMethodCodeRefactoringProvider.cs:117]: Isn't init a different type of accessor entirely. I.e. a property would have three potential accessors?  GetMethod/SetMethod/InitMethod?

Comment (jcouv) [reply src/Features/CSharp/Portable/CodeRefactorings/ConvertLocalFunctionToMethod/CSharpConvertLocalFunctionToMethodCodeRefactoringProvider.cs:117]: No, it's best thought of as a modifier on `set`. But we shortened `initonly set` to `init`.
https://github.com/dotnet/csharplang/blob/master/proposals/init.md

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/CodeRefactorings/ConvertLocalFunctionToMethod/CSharpConvertLocalFunctionToMethodCodeRefactoringProvider.cs:117]: got it 👍   can you make it a default value?  thanks!

File: src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs:137]: > What's a simple/modern test you'd recommend as a model for retargeting?

For example, look at CodeGenTests.CustomFields_01. The 7th compilation, the commented out, is supposed to use retargeting symbols for comp1. The change in the target framework does the trick. 

---
In reply to: [422651853](https://github.com/dotnet/roslyn/pull/44077#discussion_r422651853) [](ancestors = 422651853,422534751)

File: src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs [hunk 6]
```diff
@@ -13,6 +13,7 @@
 using Roslyn.Test.Utilities;
 using Xunit;
 using Utils = Microsoft.CodeAnalysis.CSharp.UnitTests.CompilationUtils;
+using Microsoft.CodeAnalysis.Test.Utilities;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs:16]: This looks like an unnecessary change, consider reverting the file.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 4)

File: src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs (hunk 2 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Retargeting/RetargetingTests.cs]: Thanks a lot. Changed the retargeting test to use that model.

**Comment by jcouv** [comment]
I've added a follow-up item to the test plan to resolve the naming question in API and spec.

File: src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs [hunk 7]
```diff
@@ -140,6 +140,12 @@ public interface IMethodSymbol : ISymbol
         /// </summary>
         bool IsReadOnly { get; }
 
+        /// <summary>
+        /// 'init' set accessors can only be invoked during construction or in the initialization phase
```
Comment (gafter) [inline src/Compilers/Core/Portable/Symbols/IMethodSymbol.cs:144]: >set [](start = 19, length = 3)

It isn't clear if this is intended to apply only to set accessors, or if other methods may be so flagged.

**Comment by jcouv** [comment]
FYI @AlekseyTs I made minor change to retargeting test to reflect different verification behavior on desktop versus Core

**Comment by jcouv** [comment]
@CyrusNajmabadi Please take another look when you can. Thanks
